Support for twig 3.9

diffy0712 opened the pull request:
| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️
| New feature?  | ❌
| Breaks BC?    | ❌
| Fixed issues  | #153

Twig 3.9 introduced a few internal changes that this package relied on:

1. The internal function `twig_get_attribute` was moved and renamed to `CoreExtension::getAttribute`.
See discussion about moving functions to internal in twig:3.9 [here](https://github.com/twigphp/Twig/issues/4040).
2. The internals of Twig does not store the html content in the output buffer anymore (but it still reads from it). This means when this package called the View::endPage function to replace the placeholders it failed to do so, because the ob was empty.

diffy0712 commented:
This patch broke one test case so I still have to look at that. 

After that I can send another pr for the 2.5 release

schmunk42 commented:
As far as I can tell your patch did not break things. It's rather another side-effect of twig 3.9, but I have no idea what exactly, the test does not find a script tag from the asset bundle.

See https://github.com/schmunk42/yii2-twig/actions/runs/8799762201 - which just has this additional commit applied https://github.com/schmunk42/yii2-twig/commit/f2bb14b48a4407c93e118e3e0f170d4db0ee1c02 (prevent installation of twig 3.9 and above) - test are passing.

I'd suggest to add this to the PR and take care about the actual problem later.

CC: @samdark @bizley @rob006

diffy0712 commented:
Yes, I have looked at it, but did not yet found the source of the problem. 
What I've found is that they introduced the 'use_yield' option (false by default), which they mention in the [changelog](https://github.com/twigphp/Twig/blob/3.x/CHANGELOG#L25C4-L29C73). The cause might come from the fact that they have rewrote some of the internals that uses ob_ functions.

It seems that the yii2 view helpers(end_body, start_body etc...) twig extensions are broken in twig3.9 for some reason. 

see commit https://github.com/twigphp/Twig/commit/3b6cbf98d

I agree, that first there should be a conflicting composer patch, so no problematic updates happens.

diffy0712 commented:
This pr already contains a partial fix for 3.9. Do you think there should be another pr for just the conflict or use this and open another with the later actual fix?

schmunk42 commented:
> This pr already contains a partial fix for 3.9. Do you think there should be another pr for just the conflict or use this and open another with the later actual fix?

See https://github.com/yiisoft/yii2-twig/pull/155

diffy0712 commented:
It seems that yii2 view helper functions like 'begin_body' 'begin_page' etc. will mark the parts of the page with '<![CDATA[YII-BLOCK-BODY-END]]>' comments and in the last 'end_page' view method call will read 'ob_get_clean()' and replace these blocks with the tags that needs to be placed there. (in case of the failing test it is 'jquery.js' file.

For some reason this does not happen with twig 3.9.
I see that the `ob_get_clean` returns an empty string instead of the html of the view with the inserted tags.

eluhr commented:
That's the same thing I found. At some point the output buffer is emptied where it shouldn't be. At least that would be my first guess.

diffy0712 commented:
I think the ob clear happens here.

https://github.com/twigphp/Twig/blob/3.x/src/Template.php#L362
and since the function does not echo anything to the buffer it will stay empty.

eluhr commented:
Yup, looks like that‘s the spot

diffy0712 commented:
It seems that they now only capture and clear whats in the ob but not echoing anymore, so it seems to me that we are unable to change the html in the buffer.

eluhr commented:
If I'm interpreting the code correctly, the twig guys are turning everything completely inside out, at least when it comes to the compilation of twig.

An error is explicitly triggered here if “echo” is used:
- https://github.com/twigphp/Twig/blob/3.x/src/Compiler.php#L78
- https://github.com/twigphp/Twig/blob/3.x/src/Compiler.php#L103

I don't see an easy way to get this running with twig version 3.9.0 and above without having to touch the view component.

diffy0712 commented:
If I understood correctly, they are preparing the code for v4 when it will only use the new yield with no ob buffer.

Maybe if we could  somehow access and modify the twig content from the TwigFunction, then we could use ob inside the viewHelper to capture the yii echo and pass it to the content of this.

Or create a custom View as u said, but I think that would be a bit inconvenient and breaking change.

eluhr commented:
I'm currently trying out exactly what you're suggesting.

and as you say, I wouldn't see a custom view component as an option either

eluhr commented:
That is my current approach. I try to use the existing `viewHelper` method and the `EVENT_AFTER_RENDER` event of the view component to get the rendered output and write it to the output buffer:

```php
public function viewHelper($context, $name = null)
{
    if ($name !== null && isset($context['this'])) {
        if ($name == 'end_page') {
            // tapping into the after render event of the view component to load the rendered content into the output buffer
            Event::on($context['this']::class, View::EVENT_AFTER_RENDER, function ($event) {
                // TODO: Is loading stuff in output buffer the way?
                // $output = $event->output;
            });
        }
        $this->call($context['this'], Inflector::variablize($name));
    }
}
```

diffy0712 commented:
Just to sum up the problem that is braking the test (as I understand it)
This package until now, relies on the fact that the twig engine stored the output html in the output buffer (ob), so the yii2 view could echo into it and could modify the whole content (replace all the yii placeholder comments with assets and stuff).
but since twig 3.9 this internal behaviour changed, and they now store the content in a new internal variable (uses yielding). https://github.com/twigphp/Twig/blob/4a7de4ab5603ad8f199222cae3951fc271a8c45d/src/Template.php#L366

So this package was modifiing twig's global variable(the ob), which it has no access to modify(only append using echo) since twig:3.9.

diffy0712 commented:
> I can think of another solution(to work without yield and use the current yii views) As far as I know, the only yii view method that actually modifies the html content (in the ob) is the end_body, which should basically always happen at the end of the twig template if it is called ofc. So if we add a flag 'end_body_enabled', and instead of running the endBody view method in the twig function once it is called, we could just enable this flag, so when the ViewRenderer render (
> 
> https://github.com/yiisoft/yii2-twig/blob/0ed174ac79326ffa4101c35a429725bd082450bb/src/ViewRenderer.php#L179
> ) method renders the page 'using twig' as alway, but with the fleg enabled, capture the variable, put it in ob and call View::endPage so it can replace the placeholders and return that from the render method.
> 

This seems to be working correctly! I will push the changes soon.

diffy0712 commented:
@schmunk42 @eluhr  I think this workaround should work with twig 3.9 version as well! Could you please check it out?

I think the best solution would be to support use_yield as it will be the only option in twig 4.0, but still I do not see any good solution for it without a breaking change in this library.

diffy0712 commented on src/TwigVersionHelper.php:
```diff
@@ -0,0 +1,29 @@
+     */
+    public static function above39(): bool
+    {
+        return !function_exists('twig_get_attribute');
```
I do not think this is the best way to check the version. Do anyone have a better idea?

eluhr replied:
Can't think of a better way to check this and since the non-existence of the function is the source of our current problems.

eluhr commented:
@diffy0712 with your current patch i am getting the following error: `ob_clean(): Failed to delete buffer. No buffer to delete` wrapping the [ob_clean call in the ViewRenderer](https://github.com/diffy0712/yii2-twig/blob/fix/twig-3.9-twig_get_attribute/src/ViewRenderer.php#L208) into an if condition like this does the trick (for me):

```php
  if ($content !== false) {
      ob_clean();
  }
```

But even after that fix i get an `HeadersAlreadySentException ` which is caused by the `echo $content` in the [ViewRenderer](https://github.com/diffy0712/yii2-twig/blob/fix/twig-3.9-twig_get_attribute/src/ViewRenderer.php#L205). But maybe thats a problem from my current setup.

What is your way for recreating the bug? I tested it by using using the layout.twig from the tests as my web application layout

```php
 'layout' => '@vendor/yiisoft/yii2-twig/tests/views/layout.twig'
```

schmunk42 commented:
@yiisoft/core-developers Given these esoteric issues, I'd even think about releasing a new major version of `yii2-twig` and require `^3.9.0`.

Because twig can be configured with several plugins, custom functions, probably usage of "protected" functions in your templates and so on. It looks highly likely to me that upgrading to `twig/twig:3.9` will raise a lot of errors in existing applications.

diffy0712 commented:
> @diffy0712 with your current patch i am getting the following error: `ob_clean(): Failed to delete buffer. No buffer to delete` wrapping the [ob_clean call in the ViewRenderer](https://github.com/diffy0712/yii2-twig/blob/fix/twig-3.9-twig_get_attribute/src/ViewRenderer.php#L208) into an if condition like this does the trick (for me):
> 
> ```
>   if ($content !== false) {
>       ob_clean();
>   }
> ```
> 
> But even after that fix i get an `HeadersAlreadySentException ` which is caused by the `echo $content` in the [ViewRenderer](https://github.com/diffy0712/yii2-twig/blob/fix/twig-3.9-twig_get_attribute/src/ViewRenderer.php#L205). But maybe thats a problem from my current setup.
> 
> What is your way for recreating the bug? I tested it by using using the layout.twig from the tests as my web application layout
> 
> ```
>  'layout' => '@vendor/yiisoft/yii2-twig/tests/views/layout.twig'
> ```

I've done the fix by running the unit tests. It seems that I have to check out in my project as well, because it seems that there is some problems with the buffer part of the code.

diffy0712 commented:
> @yiisoft/core-developers Given these esoteric issues, I'd even think about releasing a new major version of `yii2-twig` and require `^3.9.0`.
> 
> Because twig can be configured with several plugins, custom functions, probably usage of "protected" functions in your templates and so on. It looks highly likely to me that upgrading to `twig/twig:3.9` will raise a lot of errors in existing applications.

Yes that is true, the same issue that the end_page twig function failed with twig:3.9 it can happen in anyone's project using yii2-twig. 
Releasing as a major version it would at least "warn" people when upgrading, and it should be stated in the Changelog as well with detailed explanation on the possible issues. 

In this case I would remove the old version support from this pr and update the compose.json as well.

diffy0712 commented:
@eluhr I've fixed the render output buffer issue. Could you please check if it is working for you as well?
It was an issue that the $view->endPage(); method did close the output buffer, which caused a problem. I have tested it with the unit tests, which apparently has another output buffer layer started so it caused some problem :D 
Hope it is working correctly now.

diffy0712 commented:
@schmunk42 I've removed the support for twig below 3.9 and added the updated composer version dependency for twig to 3.9.
Do you think that is enough or should I add something else?

diffy0712 commented:
I still need to add description to the Changelog for the new release to explain this change and the situation with > twig 3.9 better. I will try to do it tomorrow.

eluhr commented:
> @eluhr I've fixed the render output buffer issue. Could you please check if it is working for you as well?

@diffy0712 LGTM!

schmunk42 commented:
I also ran our phd5 tests against this patch - all green :+1: 

So I'd revise my statement about a new major version :)
But at least there needs to be a new minor version with information about the deprecation of internal twig functions.

samdark commented:
@schmunk42 so do we need to release from master before merging this PR?

schmunk42 commented:
Yes, I’d recommend to release current master as 2.4.3And the PR as 2.5.0 or 3.0.0Von meinem iPhone gesendetAm 26.04.2024 um 18:08 schrieb Alexander Makarov ***@***.***>:﻿
@schmunk42 so do we need to release from master before merging this PR?

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

diffy0712 commented:
I've updated the changelog as well. I think this pr is ready for review.

schmunk42 commented:
> @schmunk42 how to continue with this?

I think it should be merged and released as `2.5.0`.

samdark commented:
@schmunk42 done. Thanks, @diffy0712
